pictorus-internal/src/loggers/csv_logger.rs

```diff
@@ -14,33 +14,123 @@ use super::Logger;
+                let mut first = true;
+                for (key, _) in json_map {
+                    if !first {
+                        self.buffer.push(',');
+                    }
+                    self.buffer.push_str(key);
+                    first = false;
+                }
+                writeln!(self.writer, "{}", self.buffer).ok();
+                self.header_written = true;
+                self.buffer.clear();
+            }
+
+            // Write values
+            let mut first = true;
+            for (_, value) in json_map {
+                if !first {
+                    self.buffer.push(',');
+                }
+
+                match value {
+                    serde_json::Value::Null => {}
+                    serde_json::Value::Bool(b) => {
+                        self.buffer.push_str(if *b { "true" } else { "false" });
+                    }
+                    serde_json::Value::Number(n) => {
+                        use std::fmt::Write;
+                        write!(self.buffer, "{n}").ok();
+                    }
+                    serde_json::Value::String(s) => {
+                        self.buffer.push('"');
+                        self.buffer.push_str(s);
+                        self.buffer.push('"');
+                    }
+                    serde_json::Value::Array(arr) => {
+                        self.buffer.push('"');
+                        self.buffer.push('[');
+                        let mut first_elem = true;
+                        for elem in arr {
+                            if !first_elem {
+                                self.buffer.push(',');
+                            }
+                            if let serde_json::Value::Number(n) = elem {
+                                use std::fmt::Write;
+                                write!(self.buffer, "{n}").ok();
+                            } else if let serde_json::Value::Array(inner) = elem {
+                                // Handle nested arrays (like [[1,2,3]])
+                                self.buffer.push('[');
+                                let mut first_inner = true;
+                                for inner_elem in inner {
+                                    if !first_inner {
+                                        self.buffer.push(',');
+                                    }
+                                    if let serde_json::Value::Number(n) = inner_elem {
+                                        use std::fmt::Write;
+                                        write!(self.buffer, "{n}").ok();
+                                    }
+                                    first_inner = false;
+                                }
+                                self.buffer.push(']');
+                            }
+                            first_elem = false;
+                        }
+                        self.buffer.push(']');
+                        self.buffer.push('"');
```

Comment:
Are we sure this is actually faster? I think the `serde_json::to_value` step has already done a lot of the heavy lifting at this point, so I'd be surprised if our own custom JSON serializer was actually much of an improvement